src/main/java/io/supertokens/totp/Totp.java

```diff
@@ -0,0 +1,318 @@
+                return false;
+            }
+        }
+
+        return false;
+    }
+
+    public static TOTPDevice registerDevice(Main main, String userId, String deviceName, int skew, int period)
+            throws StorageQueryException, DeviceAlreadyExistsException, NoSuchAlgorithmException {
+
+        TOTPSQLStorage totpStorage = StorageLayer.getTOTPStorage(main);
+
+        // TODO: There should be a hard limit on number of devices per user
+        // 8 devices per user should be enough. Otherwise, it is a security risk.
+
+        String secret = generateSecret();
+        TOTPDevice device = new TOTPDevice(userId, deviceName, secret, period, skew, false);
+        totpStorage.createDevice(device);
+
+        return device;
+    }
+
+    private static void checkAndStoreCode(Main main, TOTPStorage totpStorage, String userId, TOTPDevice[] devices,
+            String code)
+            throws InvalidTotpException, StorageQueryException, TotpNotEnabledException, LimitReachedException {
+        // Note that the TOTP cron runs every 1 hour, so all the expired tokens can stay
+        // in the db for at max 1 hour after expiry.
+
+        // If we filter expired codes in rate limiting logic, then because of
+        // differences in expiry time of different codes, we might end up with a
+        // situation where:
+        // Case 1: users might get released from the rate limiting too early because of
+        // some invalid codes in the checking window were expired.
+        // Case 2: users might face random rate limiting because if some valid codes
+        // expire and if it leads to N contagious invalid
+        // codes, then the user will be rate limited for no reason.
+
+        // For examaple, assume 0 means expired; 1 means non-expired:
+        // Also, assume that totp_max_attempts is 3, totp_rate_limit_cooldown_time is
+        // 15 minutes, and totp_invalid_code_expiry is 5 minutes.
+
+        // Example for Case 1:
+        // User is rate limited and the used codes are like this: [1, 1, 0, 0, 0]. Now
+        // if 1st zero (invalid code) expires in 5 mins and we filter
+        // out expired codes, we'll end up [1, 1, 0, 0]. This doesn't contain 3
+        // contiguous invalid codes, so the user will be released from rate limiting in
+        // 5 minutes instead of 15 minutes. 
+
+        // Example for Case 2:
+        // User has used codes like this: [0, 1, 0, 0].
+        // The 1st one (valid code) will expire in merely 1.5 minutes (assuming skew = 2
+        // and period = 30s). So now if we filter out expired codes, we'll see
+        // [0, 0, 0] and this contains 3 contagious invalid codes, so now the user will
+        // be rate limited for no reason.
+
+        // That's why we need to fetch all the codes (expired + non-expired).
+        TOTPUsedCode[] usedCodes = totpStorage.getAllUsedCodesDescOrder(userId);
```

Comment:
Yes. So what happens is if in txn1 I did `SELECT * FROM totp_used_codes WHERE user_id = <user-id> FOR UPDATE;`, and txn2 tries to do the same, it will wait until txn1 is commited/rolled back.